umbra/umbra_x86.c

```diff
@@ -295,11 +300,78 @@ shadow_table_insert_app_to_shadow(void *drcontext,
                                                              sizeof(ptr_int_t),
                                                              disp,
                                                              OPSZ_PTR)));
+}
 #elif defined(ARM)
-    /* FIXME i#1726: port to ARM by adding more XINST_CREATE_ routines */
-#endif
+int
+umbra_num_scratch_regs_for_translation_arch(void)
+{
+    return 2;
 }
 
+/* code sequence:
+ * %reg_idx   = table
+ * %reg_tmp   = (%reg_addr >> 0x10)
+ * %reg_idx  += (%reg_tmp  << 0x02)
```

Comment:
You could get rid of reg_tmp by adding (%reg_addr >> 14) here and then zeroing the bottom 2 bits, right?

Reply:
Sorry about the delay. I think this would only work if the static table already has the bottom 2 bits cleared:

```
reg_idx  =  0x41414141
reg_idx += (reg_addr >> 14)
reg_idx &= 0xfffffffc
```

If we simply clear the last two bits of `reg_idx`, we might kill 2 bits in the address table. I guess it's ok to assume that the static table is 4 byte aligned though, so maybe this is fine.

Reply:
Unrelated: there's a confusing variable called `static_shadow_table_is_used` which is statically set to false, though really the variable should be named `static_shadow_table_unused` or sthg. Can update this.

Reply:
Hmm yeah that seems to never be written to?

Reply:
Yup looks like it's only being used once to determine whether to allocate the shadow table using `nonheap_alloc()` (if `static_shadow_table_is_used` is *true*), but uses the static shadow table if `static_shadow_table_is_used` is *false*, hence the confusion.